### PR TITLE
[Keyboard] Add kudox game keyboard

### PR DESCRIPTION
**Pull request by x1-**
## Description

Adds support for Kudox Game Keyboard by [Kumao Kobo](https://github.com/kumaokobo).

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Adds support for Kudox Game Keyboard.

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/kudox_game/keymaps/default/keymap.c [hunk 1]
```diff
@@ -0,0 +1,51 @@
+#include QMK_KEYBOARD_H
+
+extern keymap_config_t keymap_config;
```
Comment (fauxpark) [inline keyboards/kudox_game/keymaps/default/keymap.c]:
```suggestion
```

This isn't used anywhere.

File: keyboards/kudox_game/keymaps/default/keymap.c [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+// Each layer gets a name for readability, which is then used in the keymap matrix below.
+// The underscores don't mean anything - you can have a layer called STUFF or any other name.
+// Layer names don't all need to be of the same length, obviously, and you can also skip them
+// entirely and just use numbers.
+#define _QWERTY 0
+#define _SYMB 1
+
+enum custom_keycodes {
+  QWERTY = SAFE_RANGE,
```
Comment (fauxpark) [inline keyboards/kudox_game/keymaps/default/keymap.c]: These keycodes aren't used and can be removed.

File: keyboards/kudox_game/rules.mk [hunk 3]
```diff
@@ -0,0 +1,73 @@
+# MCU name
+#MCU = at90usb1286
```
Comment (fauxpark) [inline keyboards/kudox_game/rules.mk]:
```suggestion
```

File: keyboards/kudox_game/rules.mk [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
+
+
+# Boot Section Size in *bytes*
+#   Teensy halfKay   512
+#   Teensy++ halfKay 1024
+#   Atmel DFU loader 4096
+#   LUFA bootloader  4096
+#   USBaspLoader     2048
+OPT_DEFS += -DBOOTLOADER_SIZE=4096
```
Comment (fauxpark) [inline keyboards/kudox_game/rules.mk]: If you know the bootloader for this keyboard, please replace this block with:

```make
# Bootloader selection
#   Teensy       halfkay
#   Pro Micro    caterina
#   Atmel DFU    atmel-dfu
#   LUFA DFU     lufa-dfu
#   QMK DFU      qmk-dfu
#   atmega32a    bootloadHID
BOOTLOADER = atmel-dfu
```

File: keyboards/kudox_game/rev1/rev1.h [hunk 5]
```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include "../kudox_game.h"
+
+#include "quantum.h"
+
+
+#ifdef USE_I2C
```
Comment (fauxpark) [inline keyboards/kudox_game/rev1/rev1.h]: As with the other keyboard, this probably can be removed.

**Comment by x1-** [comment]
@fauxpark 
Thank you for reviewing my codes!
I also modified the codes you pointed out.

File: keyboards/kudox_game/rev1/config.h [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+// wiring of each half
+#define MATRIX_ROW_PINS { D4, D7, E6, B4, B5 }
+#define MATRIX_COL_PINS { F5, F6, F7, B1, B3, B2, B6 }
+// #define MATRIX_COL_PINS { B6, B2, B3, B1, F7, F6, F5} //uncomment this line and comment line above if you need to reverse left-to-right key order
+
+/* COL2ROW or ROW2COL */
+#define DIODE_DIRECTION COL2ROW
+
+#define CATERINA_BOOTLOADER
```
Comment (fauxpark) [inline keyboards/kudox_game/rev1/config.h]:
```suggestion
```

File: keyboards/kudox_game/rules.mk [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality on B7 by default
+MIDI_ENABLE = no            # MIDI support (+2400 to 4200, depending on config)
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
+RGBLIGHT_ENABLE = no        # Enable WS2812 RGB underlight.
+
+SPLIT_KEYBOARD = no
```
Comment (fauxpark) [inline keyboards/kudox_game/rules.mk]:
```suggestion
```

File: keyboards/kudox_game/rev1/rules.mk [hunk 8]
```diff
@@ -0,0 +1 @@
+BACKLIGHT_ENABLE = no
```
Comment (fauxpark) [inline keyboards/kudox_game/rev1/rules.mk]:
```suggestion
```

File: keyboards/kudox_game/keymaps/default/config.h [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+/* Use I2C or Serial, not both */
+#define USE_SERIAL
```
Comment (fauxpark) [inline keyboards/kudox_game/keymaps/default/config.h]: As with the other board, this probably should all go in the keyboard config.h.

**Comment by x1-** [comment]
@fauxpark
Thank you for your fastest revewing!
I modified the codes you pointed out.

File: keyboards/kudox_game/rev1/config.h [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xFEED
+#define PRODUCT_ID      0x9696
+#define DEVICE_VER      0x0100
+#define MANUFACTURER    Kumao Kobo
+#define PRODUCT         The Kudox Game Keyboard
+#define DESCRIPTION     Custom keyboard for game
+
+/* key matrix size */
+// Rows are doubled-up
```
Comment (noroadsleft) [inline keyboards/kudox_game/rev1/config.h]:
```suggestion
```

File: keyboards/kudox_game/rev1/config.h [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+#define MATRIX_ROWS 15
```
Comment (noroadsleft) [inline keyboards/kudox_game/rev1/config.h]:
```suggestion
#define MATRIX_ROWS 5
```

File: keyboards/kudox_game/rev1/config.h [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+/* Mechanical locking support. Use KC_LCAP, KC_LNUM or KC_LSCR instead in keymap */
+#define LOCKING_SUPPORT_ENABLE
+/* Locking resynchronize hack */
+#define LOCKING_RESYNC_ENABLE
+
+/* ws2812 RGB LED */
+#define RGB_DI_PIN D3
+
+#define RGBLED_NUM 14    // Number of LEDs
```
Comment (noroadsleft) [inline keyboards/kudox_game/rev1/config.h]: This line is why your firmware isn't compiling.

The keyboard-level `config.h` file is processed before this file (which is revision-level). `keyboards/kudox_game/config.h` is defining `RGBLED_NUM` as 7, but then this line attempts to define it as 14.

```suggestion
#define RGBLED_NUM 7    // Number of LEDs
```

Also, personally I think all the RGB definitions should be here, but that matters less when there's only one revision being supported. But this line absolutely needs to be changed.

File: keyboards/kudox_game/rev1/config.h (hunk 10 above)
Comment (noroadsleft) [inline keyboards/kudox_game/rev1/config.h]:
```suggestion
```

File: keyboards/kudox_game/rev1/config.h (hunk 11 above)
Comment (noroadsleft) [inline keyboards/kudox_game/rev1/config.h]:
```suggestion
#define MATRIX_ROWS 5
```

File: keyboards/kudox_game/rev1/config.h (hunk 12 above)
Comment (noroadsleft) [inline keyboards/kudox_game/rev1/config.h]: This line is why your firmware isn't compiling.

The keyboard-level `config.h` file is processed before this file (which is revision-level). `keyboards/kudox_game/config.h` is defining `RGBLED_NUM` as 7, but then this line attempts to define it as 14.

```suggestion
#define RGBLED_NUM 7    // Number of LEDs
```

Also, personally I think all the RGB definitions should be here, but that matters less when there's only one revision being supported. But this line absolutely needs to be changed.

Comment (x1-) [reply keyboards/kudox_game/rev1/config.h]: > Also, personally I think all the RGB definitions should be here


I understand that you said.

**Comment by x1-** [comment]
@noroadsleft 
Thank you for reviewing, too!
I modified the codes you pointed out.

**Pull request merged**